baseapp/baseapp.go

```diff
@@ -696,26 +698,57 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte) (gInfo sdk.GasInfo, re
 		if app.postHandler != nil {
 			newCtx, err := app.postHandler(runMsgCtx, tx, mode == runTxModeSimulate)
 			if err != nil {
-				return gInfo, nil, nil, priority, err
+				return gInfo, nil, nil, priority, ctx, err
 			}
 
 			result.Events = append(result.Events, newCtx.EventManager().ABCIEvents()...)
 		}
 
 		if mode == runTxModeDeliver {
 			// When block gas exceeds, it'll panic and won't commit the cached store.
+			// this should be called before we charge additional fee( otherwise would
+			// defy the whole point of charging additional fee at the end)
 			consumeBlockGas()
 
 			msCache.Write()
 		}
 
-		if len(anteEvents) > 0 && (mode == runTxModeDeliver || mode == runTxModeSimulate) {
-			// append the events in the order of occurrence
-			result.Events = append(anteEvents, result.Events...)
+		// apply fee logic calls
+		feeEvents, errFromFeeInvoker := FeeInvoke(mode, app, runMsgCtx)
+		// if err from FeeInvoke then don't write to cache
+		if errFromFeeInvoker == nil {
+			msCache.Write()
```

Comment:
doesn't seem like this should be here since it only gets called in a `runTxModeDeliver` context in the base implementation

Reply:
Yes, this shouldn't be here.

Reply:
I revamped this chunk. I didn't do too well with the cherry-pick merge, and looked more closely at our changes vs what's in v0.46. Back down to only one `msCache.Write()` in there.